### PR TITLE
satellitedb/accounting: project usage optimization

### PR DESCRIPTION
**Pull request by crawter**
What: project usage query was optimized. Memory allocation in go was reduced and some db queries works faster and returns less data

Why: In some cases this query works slow. Our QA andrew should wait 4 sec until he gets project usage on his local machine. Also, this query is used often

So what have been changed:

1. 
```All_BucketStorageTally_By_ProjectId_And_BucketName_And_IntervalStart_GreaterOrEqual_And_IntervalStart_LessOrEqual_OrderBy_Desc_IntervalStart``` - 
this dbx query fetch all table columns, but we need only 4 here

```
SELECT
	bucket_storage_tallies.interval_start, 
	bucket_storage_tallies.inline,
	bucket_storage_tallies.remote,
	bucket_storage_tallies.object_count
```
2. there were a  roullupsQuery 
```
SELECT
	SUM(settled), SUM(inline), action
FROM
	bucket_bandwidth_rollups
WHERE 
	project_id = ? AND 
	interval_start >= ? AND 
	interval_start <= ?
		GROUP BY action
```

that return matrix with 2 columns of sum. than a loop in go part that makes sum of this matrix elements. I have created 1 query that calculates final sum and do it faster because of better where clause

```
SELECT 
	SUM(settled) + SUM(inline) 
FROM 
	bucket_bandwidth_rollups 
WHERE 
	project_id = ? AND 
	interval_start >= ? AND 
	interval_start <= ? AND 
	action IN (2, 3, 4);
```

Please describe the tests:
So, how i measured performance:

for each method (old and optimized) i created 2 benchmark tests.
In first test i just run whole method: db queries + golang code
In second one i left only queries, to see impact on database, without go code.

In both cases optimized method works ~1.5 times faster
And memory allocation from go part was reduced greatly
 
Please describe the performance impact: huge, on my machine

## Code Review Checklist (to be filled out by reviewer)
 - [x] NEW: Are there any Satellite database migrations? Are they forwards _and_ backwards compatible? 
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

File: satellite/satellitedb/projectaccounting.go [hunk 1]
```diff
@@ -225,67 +228,108 @@ func (db *ProjectAccounting) GetProjectTotal(ctx context.Context, projectID uuid
 	defer mon.Task()(&ctx)(&err)
 	since = timeTruncateDown(since)
 
-	storageQuery := db.db.All_BucketStorageTally_By_ProjectId_And_BucketName_And_IntervalStart_GreaterOrEqual_And_IntervalStart_LessOrEqual_OrderBy_Desc_IntervalStart
+	storageQuery := db.db.Rebind(`
```
Comment (ifraixedes) [inline satellite/satellitedb/projectaccounting.go]: Define this variable just before it's used first

File: satellite/satellitedb/projectaccounting.go [hunk 2]
```diff
@@ -225,67 +228,108 @@ func (db *ProjectAccounting) GetProjectTotal(ctx context.Context, projectID uuid
...
+			bucket_storage_tallies 
+		WHERE 
+			bucket_storage_tallies.project_id = ? AND 
+			bucket_storage_tallies.bucket_name = ? AND 
+			bucket_storage_tallies.interval_start >= ? AND 
+			bucket_storage_tallies.interval_start <= ? 
+		ORDER BY bucket_storage_tallies.interval_start DESC
+	`)
+
+	bucketsQuery := db.db.Rebind(`
```
Comment (ifraixedes) [inline satellite/satellitedb/projectaccounting.go]: Define this variable just before it's used first

File: satellite/satellitedb/projectaccounting.go [hunk 3]
```diff
@@ -225,67 +228,108 @@ func (db *ProjectAccounting) GetProjectTotal(ctx context.Context, projectID uuid
...
+	totalEgressQuery := db.db.Rebind(`
+		SELECT 
+			SUM(settled) + SUM(inline) 
+		FROM 
+			bucket_bandwidth_rollups 
+		WHERE 
+			project_id = ? AND 
+			interval_start >= ? AND 
+			interval_start <= ? AND 
+			action IN (2, 3, 4);
```
Comment (ifraixedes) [inline satellite/satellitedb/projectaccounting.go]: I would prefer to use the constants something like 

```go
fmt.Sprintf(`
		SELECT 
			SUM(settled) + SUM(inline) 
		FROM 
			bucket_bandwidth_rollups 
		WHERE 
			project_id = ? AND 
			interval_start >= ? AND 
			interval_start <= ? AND 
			action IN (%d, %d, %d);
`, pb.PieceAction_GET, pb.PieceAction_GET_AUDIT, pb.PieceAction_GET_REPAIR)
```

**Comment by crawter** [comment]
> A few minor comments.
> 
> On the other hand
> 
> > In some cases this query works slow.
> 
> I'm curious to know in which cases works slow and how you measure that these changes are an optimization compared with the previous ones

updated in description

File: satellite/satellitedb/projectaccounting.go [hunk 4]
```diff
@@ -225,67 +229,72 @@ func (db *ProjectAccounting) GetProjectTotal(ctx context.Context, projectID uuid
...
 
-	storageQuery := db.db.All_BucketStorageTally_By_ProjectId_And_BucketName_And_IntervalStart_GreaterOrEqual_And_IntervalStart_LessOrEqual_OrderBy_Desc_IntervalStart
-
-	roullupsQuery := db.db.Rebind(`SELECT SUM(settled), SUM(inline), action
-		FROM bucket_bandwidth_rollups
-		WHERE project_id = ? AND interval_start >= ? AND interval_start <= ?
-		GROUP BY action`)
-
-	rollupsRows, err := db.db.QueryContext(ctx, roullupsQuery, projectID[:], since, before)
+	totalEgress, err := db.getTotalEgress(ctx, projectID, since, before)
```
Comment (ifraixedes) [inline satellite/satellitedb/projectaccounting.go]: Move this to the line 285

File: satellite/satellitedb/projectaccounting.go [hunk 5]
```diff
@@ -270,7 +270,6 @@ func (db *ProjectAccounting) GetProjectTotal(ctx context.Context, projectID uuid
 				return nil, err
 			}
 			tally.BucketName = bucket
-			storageTallies = append(storageTallies, &tally)
```
Comment (ifraixedes) [inline satellite/satellitedb/projectaccounting.go]: I think that this is needed

**Pull request merged**